### ISSUE TITLE
"Connection reset by peer" error when vector agent is running inside an AKS cluster

### ISSUE THREAD

**Issue by hjarvard** [vectordotdev/vector#6779]
### Vector Version

```
version="0.12.1" git_version="v0.12.1" released="Fri, 12 Mar 2021 21:32:07 +0000" arch="x86_64"
```

### Vector Configuration File

Error is reproducible using default configuration file.

### Expected Behavior

Vector agent should work indefinitely.

### Actual Behavior

Vector agent stops after ~3-4 minutes after startup and prints "Connection reset by peer" error message.

### Example Data

[vector startup trace](https://gist.github.com/hjarvard/09564eb1f25c5cb0969abbde7045cde8#file-vector-startup-trace)

[vector error trace](https://gist.github.com/hjarvard/09564eb1f25c5cb0969abbde7045cde8#file-vector-error-trace)

tcpdump including tcp reset packet (sorry, I don't have more tcpdump records, api server is very chatty)
```
06:42:48.438775 IP 10.230.0.4.443 > 10.230.0.12.51864: Flags [R], seq 2150274191, win 0, length 0
06:42:48.438775 IP 10.230.0.4.443 > 10.230.0.12.51866: Flags [R], seq 3318747671, win 0, length 0
```
10.230.0.4 is an api server, 10.230.0.12 is a vector pod

### Additional Context

I'm trying to setup vector on an aks cluster. It's a single node private cluster using pre-created virtual network and subnet.
Vector is deployed as described in the documentation using kubectl command.

Cluster setup:
 - Cluster and node pool version: 1.19.7
 - Role-based access control: Disabled
 - Network type: Azure CNI
 - Private cluster: Enabled
 - HTTP application routing: Not enabled
 - Azure AD integration: Not enabled
 - There are no network policies or firewall rules applied

Vector agent starts correctly, and after ~1 minute delay it begins log collection from worker node. 
In ~3-4 minutes after starting up vector prints error message (see above) and shuts down kubernetes sink. 
At the same time I observe a tcp rst packet is being sent from api server to vector pod. My guess is that rst packet is sent by the load balancer which runs in front of api server. By default load balancer is set to 4 min idle timeout, although I'm not sure if that is the cause of rst. 

### References

- AKS issue where they fixed rst packet not sent when cluster running in private mode: [link](https://github.com/Azure/AKS/issues/1877)
- AKS issue about api server support of load balancer: [link](https://github.com/Azure/AKS/issues/1052)

**Comment by n0rm4l-real** [comment]
@hjarvard Experiencing the same issue, any workaround for this?

**Comment by hjarvard** [comment]
@n0rm4l-real Yes, I just set up vector as usual and used file source with exclude filter, like so:
```
[sources.k8s_files]
type = "file"
include = ["/var/log/containers/**/*.log"]
exclude = [
    "*azure*",
    "*ingress*",
    "*tunnelfront*",
    "*haproxy*",
    "*coredns*",
    "**/*prometheus*",
    "**/grafana/**",
    "**/kube-proxy/**",
    "**/grafana-sc-dashboard/**",
    "**/kube-prometheus-stack/**",
    "**/vector*",
]
```
I opened [an issue](https://github.com/Azure/AKS/issues/2209) in the aks repo, but didn't get any response from azure team.

**Comment by n0rm4l-real** [comment]
@hjarvard Thanks for the provided workaround.

I also had a conversation with Microsoft Support. Long story short, there is nothing we can do about it, since AKS will not allow us to modify parameters (TCP reset, Idle timeout and Session persistence) of Load Balancer that is used for K8S API.

**Comment by n0rm4l-real** [comment]
I'm still observing the same issue on AKS with latest version (Vector 0.15.0, AKS v1.19.11):

`Jul 27 06:37:23.756  WARN source{component_kind="source" component_name=kubernetes_qa_logs component_type=kubernetes_logs}: vector::internal_events::kubernetes::reflector: Http Error in invocation! Your k8s metadata may be stale. Continuing Loop. error=K8sStream { source: Reading { source: hyper::Error(Body, Os { code: 104, kind: ConnectionReset, message: "Connection reset by peer" }) } }
Jul 27 06:37:24.764 ERROR source{component_kind="source" component_name=kubernetes_qa_logs component_type=kubernetes_logs}: vector::internal_events::kubernetes::instrumenting_watcher: Internal log [Watch stream failed.] is being rate limited.
Jul 27 06:37:24.764  WARN source{component_kind="source" component_name=kubernetes_qa_logs component_type=kubernetes_logs}: vector::internal_events::kubernetes::reflector: Handling desync. error=Desync`

**Comment by jszwedko** [comment]
Thanks for the updated report @rimitchell . Can you confirm if Vector continues processing? I know there was a bug, as mentioned in this initial report, where Vector would stop if it encountered this error. This should have been fixed as of 0.21.2 though so Vector will retry and continue.

**Comment by jszwedko** [comment]
No worries! Thanks for confirming @rimitchell .

I'm going to close this since it is expected that Vector could see these errors (reset connections) during normal operation, but it will retry them. We have some upcoming work to lower the log level for these sorts of automatically retried errors to DEBUG.
